### ISSUE TITLE
Add logging support for: MAPE (and MAE, MSE, RMSE, NRMSE)

### ISSUE THREAD

**Issue by josephjaspers** [josephjaspers/blackcat_tensors#27]

**Comment by xinsuinizhuan** [comment]
You are so greate, i set epoch to 200 and 1000, the performace is better, and no -nans. But i guess no the more epoch the better, so we should have MAPE (and MAE, MSE, RMSE, NRMSE) information, so when arrive the expected accuracy then stop train auto.
Then i expecte print the MAPE (and MAE, MSE, RMSE, NRMSE) information every epoch!

**Comment by josephjaspers** [comment]
Added logging as of:
https://github.com/josephjaspers/blackcat_tensors/commit/e6fd6d546c44dbd388b25320b66dcbd2ea163a71

(The mnist_test.h has been modified to use it) 

```
	auto network = neuralnetwork(
		feedforward(system_tag, 784, 256),
		tanh(system_tag, 256),
		feedforward(system_tag, 256, 10),
		softmax(system_tag, 10),
		logging_output_layer(system_tag, 10, BC::nn::RMSE).skip_every(100)
    );
```

Untested on windows.
Travis output:
https://travis-ci.org/josephjaspers/BlackCat_Tensors/builds/574665487

**Comment by xinsuinizhuan** [comment]
when i replace the "outputlayer(system_tag, 10)" to "logging_output_layer(system_tag, 10, BC::nn::RMSE).skip_every(100)" in mnist_test_recurrent example, then compile it, it errors. But it works in mnist_test example.
![1566371511(1)](https://user-images.githubusercontent.com/40679769/63478111-eebf8200-c4ba-11e9-8133-f7c5d9341230.jpg)

**Comment by xinsuinizhuan** [comment]
In addition, this should be Host_Convolution2d.h 
![1566435913(1)](https://user-images.githubusercontent.com/40679769/63478330-064b3a80-c4bc-11e9-9944-5aeea05e1afb.jpg)

**Comment by josephjaspers** [comment]
Yes I fixed that, shouldn't have pushed that in. 
Currently building (successfully) now
https://travis-ci.org/josephjaspers/BlackCat_Tensors

**Comment by xinsuinizhuan** [comment]
you could try to change the way to log! exmaple, to translate the x and y to Mean_Absolute_Error, then calculate or others!

**Comment by xinsuinizhuan** [comment]
How about it now?

**Comment by josephjaspers** [comment]
I want to try solve the issue you are having but I will think of an alternate way to implement logging if we cannot find the solution.

**Comment by josephjaspers** [comment]
See:
https://github.com/josephjaspers/blackcat_tensors/commit/e6fd6d546c44dbd388b25320b66dcbd2ea163a71

Problem was an issue with Visual Studio Preview (16.3) edition. The stable releases (16.1 and 16.2) work correctly.

**Comment by xinsuinizhuan** [comment]
I am so sorry, a lot of time was wasted of you to solve this problem, becouse of my  Visual Studio Preview error edition! Then focuse on the others improtance function of the sulotion.

**Comment by josephjaspers** [comment]
No problem, Its good to know for the future anyway!

**Comment by xinsuinizhuan** [comment]
Could add the MAPE log?

**Comment by josephjaspers** [comment]
Added as of: 
https://github.com/josephjaspers/blackcat_tensors/commit/cf43de380da13cbc97e8f67c17f76f385695b1da

**Comment by xinsuinizhuan** [comment]
I log MPAE、 MAE and RMSE， but they are so big and not decrease progressively, so they could not reflect the accuracy of my model, then how do I rate the accuracy of my model?I divided the data into training data and test data. After the training, I used the test data to do the test, and then calculated the average MPAE of the whole test data, how should i do?
I write it, but break:
![1566801174(1)](https://user-images.githubusercontent.com/40679769/63670345-01b1b980-c80f-11e9-8b6f-26dd1a3e460f.jpg)

**Comment by xinsuinizhuan** [comment]
> 
> 
> I am unsure what you mean by `are so big and not decrease progressively`
> 
> https://travis-ci.org/josephjaspers/blackcat_tensors
> as the loss (from the travis build) is shown to be decreasing according to the example output.

I log the MAPE in mnist_test_recurrent, that log loss is big than actual, because so namy loss data greater than 1.
[MAPE：value_type epsilon_001.txt](https://github.com/josephjaspers/blackcat_tensors/files/3544585/MAPE.value_type.epsilon_001.txt)

In addtion, i attention to in MAPE, value_type epsilon = .001， when i make the input to [0, 1], it is so big, so i attept to set value_type epsilon = 1e-6, but the MAPE log data so biger.
[MAPE：value_type epsilon_1e-6.txt](https://github.com/josephjaspers/blackcat_tensors/files/3544586/MAPE.value_type.epsilon_1e-6.txt)

In addtion, i don't know why, the log loss is so big, but the test data set's average loss is so good performance?

**Comment by xinsuinizhuan** [comment]
> 
> 
> The `img` variable is reshabed to a cube. (It is simply for `stdout`) and it is not actually fed into the network. The `inputs` are fed into the network.
> 
> The "testing..." section is a batched operation.
> To calculate the testing output error simply do:
> 
> ```
> 	double error = BC::nn::MAPE(hyps, outputs[0]);
> 	double average_error = error / batch_size;
> 	std::cout << "Average loss: " << error << std::endl;
> ```
> 
> Which template class operations are you referring to?
> (If there is something you need to know/confused I will write documentation for it if it doesn't exist yet).

It is my falut to operate the inputs, it si should be hyps.

**Comment by xinsuinizhuan** [comment]
> 
> 
> the MAPE's error appears large when it predicts '0' when the correct answer is '1'
> 
> IE
> When expected == 1
> and epsilon is .0001
> 
> Error is: 1 / .0001 == 10000
> 
> So it is expected that MAPES may appear to have very large error at times.
> 
> A better way to log error would be categorical cross entropy loss.
> Though I have not implemented it yet.

oh ,i know.But i think it shoul be most of the MAPE is not so big, only few of them big.

**Comment by xinsuinizhuan** [comment]
In addtion, i don't know why, the log loss is so big, but the test data set's average loss is so good performance?

**Comment by josephjaspers** [comment]
MAPE is calculated via...
sum(abs((actual - expected)/(actual))) / batch_size

Is this what you expect or a different function?

**Comment by xinsuinizhuan** [comment]
> 
> 
> MAPE is calculated via...
> sum(abs((actual - expected)/(actual))) / batch_size
> 
> Is this what you expect or a different function?

Yes, it is my expect function, but i am only puzzle, why the log loss is so big, but the test data set's average loss is so good performance.

**Comment by xinsuinizhuan** [comment]
Now, when i print the average loss, to conver the BC::nn::MAPE(hyps, outputs[0]) to double is error:

        double error = BC::nn::MAPE(hyps, outputs[0]);
	double average_error = error / batch_size;
	std::cout << "Average loss: " << error << std::endl;

compile it, it errors.

**Comment by josephjaspers** [comment]
> Now, when i print the average loss, to conver the BC::nn::MAPE(hyps, outputs[0]) to double is error:
> 
> ```
>     double error = BC::nn::MAPE(hyps, outputs[0]);
> double average_error = error / batch_size;
> std::cout << "Average loss: " << error << std::endl;
> ```
> 
> compile it, it errors.

Yes, that is expected I changed the implementation of sum to something faster.
You can `(value_sum(tensor))` to use the old implementation.

**Comment by josephjaspers** [comment]
Closing as this feature has been added.
